client/src/components/Workflow/Editor/Actions/workflowActions.ts

```diff
@@ -115,11 +138,286 @@ export class CopyIntoWorkflowAction extends UndoRedoAction {
+
+    constructor(commentStore: WorkflowCommentStore, stateStore: WorkflowStateStore) {
+        super();
+
+        this.commentStore = commentStore;
+        this.stateStore = stateStore;
+
+        this.selectionState = {
+            comments: [...this.commentStore.multiSelectedCommentIds],
+            steps: [...this.stateStore.multiSelectedStepIds],
+        };
+    }
+
+    get name() {
+        return "clear selection";
+    }
+
+    run() {
+        this.commentStore.clearMultiSelectedComments();
+        this.stateStore.clearStepMultiSelection();
+    }
+
+    undo() {
+        this.selectionState.comments.forEach((id) => this.commentStore.setCommentMultiSelected(id, true));
+        this.selectionState.steps.forEach((id) => this.stateStore.setStepMultiSelected(id, true));
+    }
+}
+
+abstract class ModifySelectionAction extends UndoRedoAction {
+    commentStore;
+    stateStore;
+    selection;
+    abstract addToSelection: boolean;
+
+    constructor(commentStore: WorkflowCommentStore, stateStore: WorkflowStateStore, selection: SelectionState) {
+        super();
+
+        this.commentStore = commentStore;
+        this.stateStore = stateStore;
+        this.selection = selection;
+    }
+
+    run() {
+        this.selection.comments.forEach((id) => this.commentStore.setCommentMultiSelected(id, this.addToSelection));
+        this.selection.steps.forEach((id) => this.stateStore.setStepMultiSelected(id, this.addToSelection));
+    }
+
+    undo() {
+        this.selection.comments.forEach((id) => this.commentStore.setCommentMultiSelected(id, !this.addToSelection));
+        this.selection.steps.forEach((id) => this.stateStore.setStepMultiSelected(id, !this.addToSelection));
+    }
+}
+
+export class AddToSelectionAction extends ModifySelectionAction {
+    addToSelection = true;
+
+    get name() {
+        const combinedCountName = getCombinedCountName(this.selection.steps.length, this.selection.comments.length);
+        return `add ${combinedCountName} to selection`;
+    }
+}
+
+export class RemoveFromSelectionAction extends ModifySelectionAction {
+    addToSelection = false;
+
+    get name() {
+        const combinedCountName = getCombinedCountName(this.selection.steps.length, this.selection.comments.length);
+        return `remove ${combinedCountName} from selection`;
+    }
+}
+
+export class DuplicateSelectionAction extends CopyIntoWorkflowAction {
+    constructor(workflowId: string) {
+        const stateStore = useWorkflowStateStore(workflowId);
+        const commentStore = useWorkflowCommentStore(workflowId);
+        const stepStore = useWorkflowStepStore(workflowId);
+
+        const commentIds = [...commentStore.multiSelectedCommentIds];
+        const stepIds = [...stateStore.multiSelectedStepIds];
+
+        const comments = commentIds.map((id) =>
+            structuredClone(ensureDefined(commentStore.commentsRecord[id]))
+        ) as WorkflowComment[];
+
+        const steps = Object.fromEntries(
```

Comment:
The single item copy erases the label, as duplicate step labels are not allowed. We should probably do that here as well, as it will be even harder to figure out all the labels that need to be changed ... or highlight the (now) invalid step labels somehow.

Reply:
thank you, that's a good idea!